icu4c/source/common/cjmlbe.cpp

```diff
@@ -0,0 +1,457 @@
+// © 2022 and later: Unicode, Inc. and others.
+// License & terms of use: http://www.unicode.org/copyright.html
+
+#include <utility>
+
+#include "unicode/utypes.h"
+
+#if !UCONFIG_NO_BREAK_ITERATION
+
+#include "cjmlbe.h"
+#include "brkeng.h"
+#include "dictbe.h"
+#include "unicode/uniset.h"
+#include "unicode/chariter.h"
+#include "unicode/resbund.h"
+#include "unicode/ubrk.h"
+#include "unicode/usetiter.h"
+#include "ubrkimpl.h"
+#include "utracimp.h"
+#include "uvectr32.h"
+#include "uvector.h"
+#include "uassert.h"
+#include "unicode/normlzr.h"
+#include "cmemory.h"
+#include "dictionarydata.h"
+#include "util.h"
+
+U_NAMESPACE_BEGIN
+
+#if !UCONFIG_NO_NORMALIZATION
+
+class Element : public UMemory {
+public:
+    /**
+     * Default constructor.
+     *
+     * @param ch A Unicode character
+     * @param index The unicode block index of the character.
+     */
+    Element(UChar32 character, UnicodeString idx) : ch(character), index(idx) {}
+
+    UChar32 ch;
+    UnicodeString index;
+};
+
+CjMLBreakEngine::CjMLBreakEngine(const UnicodeSet &digitOrOpenPunctuationOrAlphabetSet,
+                                 const UnicodeSet &closePunctuationSet, UErrorCode& status)
+    : fDigitOrOpenPunctuationOrAlphabetSet(digitOrOpenPunctuationOrAlphabetSet),
+      fClosePunctuationSet(closePunctuationSet) {
+    if (U_FAILURE(status)) {
+        return;
+    }
+
+    fUnicodeBlock = new UVector32(status);
+    loadMLmodel(status);
+}
+
+CjMLBreakEngine::~CjMLBreakEngine() {
+    fUnicodeBlock->removeAllElements();
+    delete fUnicodeBlock;
+}
+
+namespace {
+    const char16_t INVALID = u'|';
+}
+
+static UnicodeString* concatChar(const char16_t *str, const UChar32 *arr, UErrorCode &status) {
```

Comment:
for all the case you use concatChar, you know the valid length in arr, why not pass in that length instead of depending on the checking of -1 in arr to have a hacky -1 terminator ? BTW, what are stored in these UnicodeString? Are those real "TEXT"? or are you trying to use it as a array of 32bit int?

Reply:
1. Add length.
2. The UnicodeString is TEXT, used to compare with ML model's feature. It's better use this object rather than primitive int array.